### PR TITLE
Use random number generators, not global seeds

### LINKED ISSUE #52: Use numpy random Generator for random numbers
### What should we add?

Changing the global random state like this is discouraged:
https://github.com/Qiskit/qiskit-addon-sqd/blob/7ddfb2101c7f2cefb75f06340dfbf42f5c6cc080/qiskit_addon_sqd/configuration_recovery.py#L213

Instead, a Generator object should be initialized using `np.random.default_rng(seed)` and used.

### PR DESCRIPTION
**Pull request by caleb-johnson**
Fixes #52 

This PR removes the use of global seeds to control numpy randomness throughout the source

### CODE REVIEW

**Comment by caleb-johnson** [comment]
Actually the tutorial experiment no longer converges after these changes. Need to look further into it

**Comment by caleb-johnson** [comment]
> Actually the tutorial experiment no longer converges after these changes. Need to look further into it

False alarm, the new generator was just giving a particularly bad sampling with that seed so config recovery couldn't bootstrap. Changing the seed brought things back to normal

File: qiskit_addon_sqd/subsampling.py [hunk 1]
```diff
@@ -136,7 +135,8 @@ def subsample(
     batches = []
     for _ in range(num_batches):
         if randomly_sample:
-            indices = np.random.choice(
+            rng = np.random.default_rng(rand_seed)
```
Comment (kevinsung) [inline qiskit_addon_sqd/subsampling.py:138]: The `rng` should always be initialized only once, at the beginning of the function. If it's initialized more than once in the function (like here), that's a bug.

**Review by kevinsung — CHANGES_REQUESTED** [review]
The notebooks need to be updated too, to use a single random generator rather than repeatedly passing the same integer seed.

**Comment by caleb-johnson** [comment]
> The notebooks need to be updated too, to use a single random generator rather than repeatedly passing the same integer seed.

So this means all of the functions need to be updated to accept a `Generator`?

**Comment by kevinsung** [comment]
> > The notebooks need to be updated too, to use a single random generator rather than repeatedly passing the same integer seed.
> 
> So this means all of the functions need to be updated to accept a `Generator`? We can deprecate the usage of `int` for `rand_seed` and move to this. I just want to make sure I understand best practices

Any function that accepts a `seed` should immediately pass the seed to `np.random.default_rng`. So the type of `seed` is the type of a valid input of `np.random.default_rng`. I never quite figured out the correct type annotation for that, but I think `int | np.random.Generator | None` will do.

File: docs/how_tos/choose_subspace_dimension.ipynb [hunk 2]
```diff
@@ -65,10 +65,11 @@
    "metadata": {},
    "outputs": [],
    "source": [
+    "import numpy as np\n",
     "from qiskit_addon_sqd.counts import generate_counts_uniform\n",
     "\n",
     "# Create a seed to control randomness throughout this workflow\n",
-    "rand_seed = 42\n",
+    "rand_seed = np.random.default_rng(2**24)\n",
```
[the remark this answers is not in the document]
Comment (caleb-johnson) [reply docs/how_tos/choose_subspace_dimension.ipynb:72]: The docs say to [use a very large number](https://numpy.org/doc/stable/reference/random/index.html). I figure this is no worse than 42

[the remark this answers is not in the document]
Comment (caleb-johnson) [reply docs/how_tos/choose_subspace_dimension.ipynb:72]: Seeds should be large positive integers. [default_rng](https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.default_rng) can take positive integers of any size. We recommend using very large, unique numbers to ensure that your seed is different from anyone else’s.

File: qiskit_addon_sqd/configuration_recovery.py [hunk 3]
```diff
@@ -92,6 +92,9 @@ def recover_configurations(
              arXiv:2405.05068 [quant-ph].
 
     """
+    if isinstance(rand_seed, int) or rand_seed is None:
+        rand_seed = np.random.default_rng(rand_seed)
```
[the remark this answers is not in the document]
Comment (caleb-johnson) [reply qiskit_addon_sqd/configuration_recovery.py]: > There's no need for the if statement. If the input is a Generator, `np.random.default_rng` will return it unchanged. Also, please be consistent with the variable name. I suggest using `rng` for the Generator everywhere.

done, thanks

File: docs/how_tos/choose_subspace_dimension.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (caleb-johnson) [reply docs/how_tos/choose_subspace_dimension.ipynb]: Updated the varnames in notebooks

[the remark this answers is not in the document]
Comment (kevinsung) [reply docs/how_tos/choose_subspace_dimension.ipynb]: The motivation for using a large number is to ensure that you don't accidentally choose the same seed as somebody else in the world. But this is not a concern here, where we only need our results to be statistically valid by themselves. Usually for these situations (testing, or just to make a notebook deterministic), you can use any integer, even 0.

But if you insist on using a very large number, then certainly `2**24` is not big enough. The doc you link suggests this:
```python
secrets.randbits(128)  
```
But again, this is really unnecessary here.

[the remark this answers is not in the document]
Comment (caleb-johnson) [reply docs/how_tos/choose_subspace_dimension.ipynb]: It truly doesn't matter at all to me. I would think 42 (a number used by everyone in every context for everything) would be worse. I will use 128 👍

[the remark this answers is not in the document]
Comment (kevinsung) [reply docs/how_tos/choose_subspace_dimension.ipynb]: Thanks. In this situation, 42 is fine. If we were going to publish scientific results, then we should indeed use a large number, such as 266529787016279213832423171828752297713.

[the remark this answers is not in the document]
Comment (caleb-johnson) [reply docs/how_tos/choose_subspace_dimension.ipynb]: OK I updated them all to ``24`` 🤝

**Review by kevinsung — APPROVED** [review]

**Pull request merged**